common/setups/rasr/nn_system.py

```diff
@@ -0,0 +1,564 @@
+__all__ = ["NnArgs", "NnSystem"]
+
+import copy
+import itertools
+import sys
+from typing import Dict, List, Optional, Tuple, Union
+
+# -------------------- Sisyphus --------------------
+
+import sisyphus.toolkit as tk
+import sisyphus.global_settings as gs
+
+# -------------------- Recipes --------------------
+
+import i6_core.rasr as rasr
+import i6_core.returnn as returnn
+
+from .rasr_system import RasrSystem
+
+from .util import (
+    RasrInitArgs,
+    ReturnnRasrDataInput,
+    OggZipHdfDataInput,
+    NnArgs,
+    RasrSteps,
+)
+
+# -------------------- Init --------------------
+
+Path = tk.setup_path(__package__)
+
+# -------------------- System --------------------
+
+
+class NnSystem(RasrSystem):
+    """
+    - 5 corpora types: train, devtrain, cv, dev and test
+        devtrain is a small split from the train set which is evaluated like
+        the cv but not used for error calculating. Since we can have different
+        datasubsets per subepoch, we do not caculate the tran score/error on
+        a consistent datasubset
+    - two training data settings: defined in returnn config or not
+    - 3 different types of decoding: returnn, rasr, rasr-label-sync
+    - 2 different lm: count, neural
+    - cv is dev for returnn training
+    - dev for lm param tuning
+    - test corpora for final eval
+
+    settings needed:
+    - am
+    - lm
+    - lexicon
+    - ce training
+    - ce recognition
+    - ce rescoring
+    - smbr training
+    - smbr recognition
+    - smbr rescoring
+    """
+
+    def __init__(
+        self,
+        returnn_root: Optional[str] = None,
+        returnn_python_home: Optional[str] = None,
+        returnn_python_exe: Optional[str] = None,
+    ):
+        super().__init__()
+
+        self.returnn_root = returnn_root or (
+            gs.RETURNN_ROOT if hasattr(gs, "RETURNN_ROOT") else None
+        )
+        self.returnn_python_home = returnn_python_home or (
+            gs.RETURNN_PYTHON_HOME if hasattr(gs, "RETURNN_PYTHON_HOME") else None
+        )
+        self.return_python_exe = returnn_python_exe or (
+            gs.RETURNN_PYTHON_EXE if hasattr(gs, "RETURNN_PYTHON_EXE") else None
+        )
+
+        self.tf_fwd_input_name = "tf-fwd-input"
+
+        self.cv_corpora = []
+        self.devtrain_corpora = []
+
+        self.train_input_data = None
+        self.cv_input_data = None
+        self.devtrain_input_data = None
+        self.dev_input_data = None
+        self.test_input_data = None
+
+        self.train_cv_pairing = None
+
+        self.datasets = {}
+
+        self.oggzips = {}
+
+        self.nn_configs = {}
+        self.nn_models = {}
```

Comment:
I just saw that there is a big inconsistency with respect to "Models" and "Checkpoints". We should only use one, so that all stuff like "getting best epoch", "checkpoint averaging" and so on will be working. I think you switched to `Checkpoint` in your further code, and I would suggest to do this everywhere, also here.